src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,159 @@
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return searchNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = searchNodeByIndex(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> currentNode = searchNodeByIndex(index);
+        unlink(currentNode);
+        return currentNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        int i = 0;
+        if (object == null) {
+            for (MyLinkedList.Node<T> j = head; j != null; j = j.next, i++) {
+                if (j.value == null) {
+                    unlink(j);
+                    return true;
+                }
+            }
+        } else {
+            for (MyLinkedList.Node<T> j = head; j != null; j = j.next, i++) {
+                if (object.equals(j.value)) {
+                    unlink(j);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public Node<T> searchNodeByIndex(int index) {
+        isIndexValid(index);
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
+    }
+
+    public void isIndexValid(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " is invalid");
+        }
+    }
+
+    public void addFirst(T value) {
+        Node<T> node;
+        if (isEmpty()) {
+            node = new Node<>(null, value, null);
+            head = node;
+            tail = node;
+        } else {
+            node = new Node<>(null, value, head);
+            head.prev = node;
+            head = node;
+        }
+        size++;
+    }
+
+    public void unlink(Node<T> node) {
+        if (node == head) {
+            if (size == 1) {
+                head = null;
+                tail = null;
+            } else {
+                node.next.prev = null;
+                head = node.next;
+                node.next = null;
+            }
+            size--;
+        } else if (node == tail) {
+            tail = node.prev;
+            node.prev.next = null;
+            node.prev = null;
+            size--;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+            node.prev = null;
+            node.next = null;
+            size--;
```

Comment:
can we simplify this?
for example, you're always decreasing size, you can make it after all if-else blocks

Reply:
Yes, you're right! Done!